File: sample_codes/sample.py
# paste this into the assistant to test

import os, sys, pickle
import hashlib

# hardcoded credentials
DB_PASSWORD = "admin123"
SECRET_KEY = "mysecretkey_hardcoded"

def get_user(u, p):
    import sqlite3
    conn = sqlite3.connect("users.db")
    cursor = conn.cursor()
    # SQL injection vulnerability
    query = "SELECT * FROM users WHERE username='" + u + "' AND password='" + p + "'"
    cursor.execute(query)
    return cursor.fetchone()
    # connection never closed

def hash_pw(password):
    # weak hashing
    return hashlib.md5(password.encode()).hexdigest()

def divide_numbers(a, b):
    unused_var = "this does nothing"
    result = a / b   # no zero division check
    return result

def calculate(expression):
    return eval(expression)   # code injection risk

def load_data(filepath):
    with open(filepath, "rb") as f:
        return pickle.load(f)   # insecure deserialization

def login(username, password):
    print(f"Attempting login with password: {password}")   # logging sensitive data
    user = get_user(username, password)
    if user == None:   # should use `is None`
        return False
    return True

def wait_for_condition(flag):
    while flag == False:   # busy wait / infinite loop risk
        pass

def read_file(path):
    try:
        f = open(path)
        data = f.read()
        return data
        f.close()   # unreachable code
    except:
        pass   # swallowing all exceptions

File: sample_codes/sample.java
import java.sql.*;
import java.util.Base64;
import java.io.*;

public class UserService {

    // hardcoded credentials
    private static final String DB_URL = "jdbc:mysql://localhost/db";
    private static final String DB_USER = "root";
    private static final String DB_PASS = "root1234";

    // SQL injection
    public User getUser(String username) throws SQLException {
        Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASS);
        Statement stmt = conn.createStatement();
        // never closed — resource leak
        ResultSet rs = stmt.executeQuery(
            "SELECT * FROM users WHERE username = '" + username + "'"
        );
        if (rs.next()) {
            return new User(rs.getString("username"));
        }
        return null;
    }

    // weak encoding — not encryption
    public String encodePassword(String password) {
        return Base64.getEncoder().encodeToString(password.getBytes());
    }

    // swallowing exception
    public String readConfig(String path) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(path));
            return br.readLine();
            // file never closed
        } catch (Exception e) {
            // silently ignored
            return null;
        }
    }

    // null check missing
    public String getUserEmail(User user) {
        return user.getEmail().toLowerCase();   // NPE if user or email is null
    }

    // infinite loop risk
    public void waitForFlag(boolean flag) {
        while (!flag) {
            // busy wait, no sleep, no timeout
        }
    }

    // logging sensitive data
    public void processPayment(String cardNumber, String cvv) {
        System.out.println("Processing card: " + cardNumber + " CVV: " + cvv);
    }
}